tests/tests/tier0/bluechi-list-units-on-all-nodes/test_bluechi_list_units_on_all_nodes.py

```diff
@@ -0,0 +1,110 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+import re
+
+from typing import Dict, Tuple
+
+from bluechi_test.config import BluechiControllerConfig, BluechiAgentConfig
+from bluechi_test.machine import BluechiControllerMachine, BluechiAgentMachine
+from bluechi_test.test import BluechiTest
+
+
+node_foo_name = "node-foo"
+node_bar_name = "node-bar"
+
+
+def parse_bluechictl_output(output: str) -> Dict[str, Dict[str, Tuple[str, str]]]:
+    line_pat = re.compile(r"""\s*(?P<node_name>[\S]+)\s*\|
+                              \s*(?P<unit_name>[\S]+)\s*\|
+                              \s*(?P<state>[\S]+)\s*\|
+                              \s*(?P<sub_state>[\S]+)\s*""",
+                          re.VERBOSE)
+    result = {}
+    for line in output.splitlines():
+        if line.startswith("NODE ") or line.startswith("===="):
+            # Ignore header lines
+            continue
+        match = line_pat.match(line)
+        if match:
+            node_units = result.get(match.group("node_name"))
+            if not node_units:
+                node_units = {}
+                node_units[match.group("unit_name")] = (match.group("state"), match.group("sub_state"))
+                result[match.group("node_name")] = node_units
+            elif match.group("unit_name") in node_units:
+                raise Exception(f"Error parsing bluechictl list-units output, unit already reported, line: '{line}'")
+            else:
+                node_units[match.group("unit_name")] = (match.group("state"), match.group("sub_state"))
```

Comment:
I think this could be simplified a bit, e.g.:

```suggestion
            if match.group("unit_name") in node_units:
                raise Exception(f"Error parsing bluechictl list-units output, unit already reported, line: '{line}'")
            
            if not node_units:
                node_units = {}
                result[match.group("node_name")] = node_units

            node_units[match.group("unit_name")] = (match.group("state"), match.group("sub_state"))
```

Reply:
You cannot use `in` operator for variable with `None` value (which happens for the 1st unit parsed for a node), that's why I have `if not not_units` condition first

Reply:
Hmm... how about explicitly adding the `node_units is not None` check? Otherwise the order of those two ifs can still be reversed - just to avoid having the duplicate `node_units[match.group("unit_name")] = (match.group("state"), match.group("sub_state"))`

Reply:
Done